soc/bouffalolab/bl60x/soc.c

```diff
@@ -0,0 +1,890 @@
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+
+	/* set PLL SDMIN */
+	/* Isn't this already set at boot by the rom settings
+	 * and we can query the value?
+	 */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_SDM_OFFSET);
+	switch (crystal) {
+	case 0:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x3C0000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 1:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x500000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 3:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x3C0000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 4:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x320000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 5:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x300000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 2:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x49D39D) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	case 32:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x3C0000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+
+	default:
+		tmp = (tmp & PDS_CLKPLL_SDMIN_UMSK)
+		    | ((uint32_t)(0x3C0000) << PDS_CLKPLL_SDMIN_POS);
+	break;
+	}
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_SDM_OFFSET);
+
+	/* phase comparator settings? */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_FBDV_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SEL_FB_CLK_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_SEL_FB_CLK_POS);
+	tmp = (tmp & PDS_CLKPLL_SEL_SAMPLE_CLK_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_SEL_SAMPLE_CLK_POS);
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_FBDV_OFFSET);
+
+	 /* Turn PLL back ON */
+	 /* frequency stabilization ? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_PU_CLKPLL_SFREG_UMSK)
+	    | ((uint32_t)(1) << PDS_PU_CLKPLL_SFREG_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	 /* let settle for a while (5 us in SDK), we may not be
+	  * running at 32Mhz right now
+	  */
+	system_clock_delay_32M_ms(2);
+
+	 /* enable PPL clock actual? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_PU_CLKPLL_UMSK)
+	    | ((uint32_t)(1) << PDS_PU_CLKPLL_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	/* More power up sequencing*/
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_PU_CP_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_CP_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_PFD_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_PFD_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_FBDV_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_FBDV_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_POSTDIV_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_POSTDIV_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	system_clock_delay_32M_ms(2);
+
+	/* reset couple things one by one? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDM_RESET_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_SDM_RESET_POS);
```

Comment:
Out of curiosity: is this equivalent?

```suggestion
	    | (1u << PDS_CLKPLL_SDM_RESET_POS);

```

Reply:
Yes, fixed. There are improvements on this field in the near future.